File: app/src/main/java/com/example/tetrisapp/model/Piece.java
package com.example.tetrisapp.model;

import androidx.annotation.NonNull;

import com.example.tetrisapp.util.ArrayHelper;

public class Piece {
    private int row = 0;
    private int col = 0;
    private byte[][] matrix;
    private final String name;
    private final int color;
    private final int overlayResId;

    public Piece(int row, int col, String name, byte[][] matrix, int color, int overlayResId) {
        this.row = row;
        this.col = col;
        this.name = name;
        this.matrix = matrix;
        this.color = color;
        this.overlayResId = overlayResId;
    }

    public Piece(String name, byte[][] matrix, int color, int overlayResId) {
        this.name = name;
        this.matrix = matrix;
        this.color = color;
        this.overlayResId = overlayResId;
    }

    public Piece copy() {
        return new Piece(row, col, name, ArrayHelper.deepCopy(matrix), color, overlayResId);
    }

    public int getRow() {
        return row;
    }

    public int getCol() {
        return col;
    }

    public byte[][] getMatrix() {
        return matrix;
    }

    public int getColor() {
        return color;
    }

    public int getOverlayResId() {
        return overlayResId;
    }

    public String getName() {
        return name;
    }

    public void setRow(int row) {
        this.row = row;
    }

    public void setCol(int col) {
        this.col = col;
    }

    public void setMatrix(byte[][] matrix) {
        this.matrix = matrix;
    }

    @NonNull
    @Override
    public String toString() {
        return name;
    }
}

File: app/src/main/java/com/example/tetrisapp/model/game/Piece.kt
package com.example.tetrisapp.model.game

class Piece {
    var row = 0
    var col = 0
    var matrix: Array<ByteArray>
    val name: String
    val color: Int
    val overlayResId: Int

    constructor(
        row: Int,
        col: Int,
        name: String,
        matrix: Array<ByteArray>,
        color: Int,
        overlayResId: Int
    ) {
        this.row = row
        this.col = col
        this.name = name
        this.matrix = matrix
        this.color = color
        this.overlayResId = overlayResId
    }

    constructor(name: String, matrix: Array<ByteArray>, color: Int, overlayResId: Int) {
        this.name = name
        this.matrix = matrix
        this.color = color
        this.overlayResId = overlayResId
    }

    fun copy(): Piece {
        return Piece(
            row,
            col,
            name,
            matrix.map { it.copyOf() }.toTypedArray(),
            color,
            overlayResId
        )
    }

    override fun toString(): String {
        return name
    }
}